[DepScan] -fdepscan driver support with new phase

cachemeifyoucan opened the pull request:
Put depscan action into a new phase in the Jobs so the driver can be run
with `-fdepscan` without running dependency scanning.

cachemeifyoucan committed:
[DepScan] -fdepscan driver support with new phase

Put depscan action into a new phase in the Jobs so the driver can be run
with `-fdepscan` without running dependency scanning.

cachemeifyoucan commented:
@swift-ci please test

benlangmuir commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4359,9 +4374,13 @@ void Driver::BuildJobs(Compilation &C) const {
-      J.InProcess = false;
+  if (C.getJobs().size() > 1 || CCPrintProcessStats) {
+    if (C.getJobs().size() != 2 ||
+        !isa<DepscanJobAction>(C.getJobs().begin()->getSource())) {
```
Worth a comment for why depscan gets special handling here

cachemeifyoucan replied:
Done. Basically it is not merging depscan into clang -cc1 so it will end up with two jobs but we really don't want to pay the overhead of process launch so we want to keep depscan in-process.

From testing, I don't see problem running two in-process commands while have one command pass file to another. If we found later that is not the case, we probably need more special handling for depscanning.

dexonsmith replied:
It doesn't seem like `CCPrintProcessStats` should trigger out-of-process `-cc1depscan`. Can you confirm why/whether this makes sense?

dexonsmith replied:
> It doesn't seem like `CCPrintProcessStats` should trigger out-of-process `-cc1depscan`. Can you confirm why/whether this makes sense?

Looking more closely, I think this is fully consistent with the existing code for `CCPrintProcessStats`. I think we can mark this as "resolved" again 👍 .

benlangmuir commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -4460,11 +4460,96 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+
+    // FIXME: We overwrite the output and the action kind argument. Maybe there
+    // are better way to do that. Not all output kind is handled here.
+    if (isa<AssembleJobAction>(JA)) {
```
Could we factor the code that handles the action into a function so we don't need to duplicate it?

cachemeifyoucan replied:
I added a comment here. This is to avoid horrible merge conflicts which I already hit. That should absolutely be a separate function when we upstreaming.

dexonsmith replied:
Can we just skip this? If the depscanning action gets the response file via `-fdepscan-output` (or `-o`) and then *also* receives the follow-up action's `-emit-llvm` (or whatever) and `-o` as part of `-cc1-args`, then it can include the `-o` as part of the response file.

To be clear, I'm thinking this command should be:
```
% clang -cc1 @file.rsp
```
so `-###` ends up with either:
```
% clang -cc1depscan -fdepscan-output file.rsp -fdepscan-args=... --cc1-args -emit-llvm -o file.bc
% clang -cc1 @file.rsp
```
or
```
% clang -cc1depscan -o file.rsp -fdepscan-args=... --cc1-args -emit-llvm -o file.bc
% clang -cc1 @file.rsp
```

cachemeifyoucan replied:
To do what you are suggesting, we need depscan to know the output type, which is not easy currently, because:
* The job in the front can't peek into the job later to see what they are doing
* The dependency scanning job currently has the output type set to be a response file
* The final output type is set after combining a series job into one and it would be hard to set the final output when the depscan phase is created.

I will take a look to see if we can put some into phase combine stage and maybe it can be done there.

dexonsmith replied:
Maybe you can add something to the end of `ToolSelector::getTool()`, right after the call to `combineWithPreprocessor()`, called `mixWithScanning()` or something?

I agree that it's a bit awkward. Phase combining feels like it should be an optimization, not a requirement, but in this case it'd be a requirement.

I suppose another approach (totally different from this patch) would be to add an `-fdepscan` option to `-cc1`... and `cc1_main()` is what connects to the daemon, etc.; then the `-cc1` is canonicalized and a new `CompilerInvocation` created and we compile from there. That model aligns less well with integrating with libclang / build system / etc. though.

cachemeifyoucan commented:
@benlangmuir @dexonsmith This change does come with overhead (in-process launch, commandline parsing multiple times, etc.). Looking at my data for a release build of clang with primed cache:
Before: 14.4s
After: 14.9s

Will be good if someone can verify but this is not nothing.

Edit: few more runs and it seems to be within the error margin from a ninja build. maybe it is fine?

dexonsmith commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -4685,6 +4774,9 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     } else if (JA.getType() == types::TY_RewrittenLegacyObjC) {
       CmdArgs.push_back("-rewrite-objc");
       rewriteKind = RK_Fragile;
+    } else if (JA.getType() == types::TY_CMD) {
+      // DepScan response file output. Use fsyntax-only is enough.
+      CmdArgs.push_back("-fsyntax-only");
```
Can you explain what this is for? I feel like the depscan job should be getting the "real" `-cc1` command, not something else.

dexonsmith replied:
(If we need to specify *something*, seems like `-Eonly` might be a better fit, since it's only doing preprocessing in the end... but I'm wondering if we can avoid this entirely, or pass the options for the job-being-scanned-for?)

cachemeifyoucan replied:
This is the same question as above. Probably need JA to carry the correct final output type (would be good not to lose TY_CMD type as that is used to create temp file but that could be implicit).

dexonsmith commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -4486,11 +4486,100 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+    Args.AddAllArgs(CmdArgs, DepScanOpts);
+
+    assert(Output.isFilename() && "Depscan needs to have file output");
+    CmdArgs.push_back("-o");
+    CmdArgs.push_back(Output.getFilename());
```
Can we use `-fdepscan-output` here so we don't collide with `-o`?

dexonsmith replied:
Actually, I don't think we'd even collide. Everything from after `-cc1-args` will get processed separately, so we could have:
```
-o file.rsp --cc1-args -emit-llvm -o file.bc
```
without any trouble, right?

In which case, I can't decide whether reusing `-o` or having a separate flag `-fdepscan-output` is better.

cachemeifyoucan replied:
Yes, exactly. `--cc1-args` split the command line and everything afterwards are just argument values, not flags.

dexonsmith reviewed (commented):
This is awesome!

> @benlangmuir @dexonsmith This change does come with overhead (in-process launch, commandline parsing multiple times, etc.). Looking at my data for a release build of clang with primed cache:
> Before: 14.4s
> After: 14.9s
> 
> Will be good if someone can verify but this is not nothing.
> 
> Edit: few more runs and it seems to be within the error margin from a ninja build. maybe it is fine?

Did you update the "After" estimate to match the smaller times? Or is it smaller than 0.5s now? 0.5s is 3-4% so it seems significant to me, but not necessarily a blocker.

My intuition is the problem is the extra filesystem work. Writing the response file and reading it.

I'm curious if it goes away if you do a really gross hack to pass the response file in memory and skip the filesystem operations. You can do this in `cc1_main`:
```
if (Argc == 3 && Argv[1] == StringRef("-cc1") && Argv[2][0] == '"') {
  getDepscanResponse(Argc, Argv);
}
```
in `cc1_depscan`, don't write out the response file, but do:
```
setDepscanResponse(...);
```
and then implement:
```
namespace {
struct DepscanResponse {
  BumpPtrAllocator Alloc;
  const char **Argv = nullptr;
  int Argc = 0;
};
}
ManagedStatic<DepscanResponse> HackyResponse;
void setDepscanResponse(...) {
  HackyResponse.Argc = Args.size();
  HackyResponse.Argv = HackyResponse.Alloc(sizeof(const char *) * (Args.size() + 1));
  StringSaver Saver(HackyResponse.Alloc);
  const char **NextArg = HackyResponse.Argv;
  for (StringRef Arg : Args)
   *NextArg++ = Saver(Arg).data();
  *NextArg = nullptr;
}
void getDepscanResponse(int &Argc, const char **&Argv) {
  if (!HackyResponse)
    return;

  Argc = HackyResponse->Argc;
  Argv = HackyResponse->Argv;
}
```
Probably easiest to have the driver set a global variable that says "don't write the response" and instead `-cc1depscan` will call `setDepscanResponse()`, and then `-cc1` will read that way instead of the original.

If it has positive impact, we can find the right way to layer it. If it's hard to implement, can explain in a FIXME what to do, and leave it for later (and for now, accept the 0.5s regression).

dexonsmith commented:
(Just marked a couple of Ben's comments as "unresolved" since I added more text to the same ones... maybe I should have created new comments instead!)

cachemeifyoucan commented:
@dexonsmith do you have any good setup for testing the overhead? I update the time after a second series run which averages about .5s less (which is larger than current gap). The minimal time is very comparable (around 14.1s) but the average has a bit distance to close but it could also be my system is not stabilized for that measurement.

dexonsmith commented:
> @dexonsmith do you have any good setup for testing the overhead? I update the time after a second series run which averages about .5s less (which is larger than current gap). The minimal time is very comparable (around 14.1s) but the average has a bit distance to close but it could also be my system is not stabilized for that measurement.

Minimal time is what matters in my opinion, due to stabilization. Even in LNT with stabilization we used to use best-of-5 I think. Everything else is kind of random/arbitrary.

cachemeifyoucan committed:
[DepScan] Teach clang contruct job to construct depscan job better

Teach clang::ConstructJob to create DepScan job based on the actual
compiling job.

cachemeifyoucan commented:
Added a new implementation that doesn't cause the code duplication (but might cause merge conflict if anything changes)

dexonsmith commented on clang/include/clang/Driver/Types.def:
```diff
@@ -102,3 +102,6 @@ TYPE("cuda-fatbin",              CUDA_FATBIN,  INVALID,         "fatbin", phases
 TYPE("none",                     Nothing,      INVALID,         nullptr,  phases::Compile, phases::Backend, phases::Assemble, phases::Link)
+
+// Generic command line type, used to pass command as response file to the next action.
+TYPE("command",                  CMD,          INVALID,         "cmd",    phases::Depscan)
```
I think the usual extension for response files is `.rsp`.

dexonsmith commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4723,6 +4746,14 @@ class ToolSelector final {
     Inputs = NewInputs;
   }
 
+  void combineWithDepscan(const Tool *T, const JobAction *Current,
```
I think there should be a good header comment here explaining what this does.

dexonsmith commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4779,6 +4810,8 @@ class ToolSelector final {
 
     combineWithPreprocessor(T, Inputs, CollapsedOffloadAction);
+
+    combineWithDepscan(T, BaseAction, Inputs);
```
I suggest adding a comment here saying this needs to be last, so it picks up the right action.

dexonsmith commented on llvm/include/llvm/CAS/HierarchicalTreeBuilder.h:
```diff
@@ -81,6 +81,9 @@ class HierarchicalTreeBuilder {
   Expected<TreeRef> create(CASDB &CAS);
+
+  /// Clone a tree in CASDB, return the top-level \a CASID in new CASDB.
+  static Expected<TreeRef> clone(CASDB &Src, TreeRef Root, CASDB &Dst);
```
I'm not sure this API really belongs in HierarchicalTreeBuilder. Seems unrelated to the filesystem stuff.

What makes sense to me is:
```
Expected<CASID> CASDB::cloneToForeignCAS(const CASID &ID, CASDB &DestCAS);
```
For now, a generic `CASDB` function. But eventually:
- We make it `virtual` and `override` in implementations for faster object traversal.
- If both are an on-disk CAS, we use `clone_file` for large/outlined objects.

(I'm fine if you only make it work for trees and blobs to start.)

dexonsmith replied:
Also, was it intentional to put this in the `-fdepscan` patch? It doesn't look like you call it... and it'd be MUCH nicer to have in a separate PR I think.

dexonsmith commented on llvm/lib/CAS/HierarchicalTreeBuilder.cpp:
```diff
@@ -257,3 +257,32 @@ Expected<TreeRef> HierarchicalTreeBuilder::create(CASDB &CAS) {
+          return Error::success();
+
+        // Clone the blob nodes.
+        // TODO: support complicated node/tree types?
```
Definitely should support all objects. Not just trees as inputs.

dexonsmith commented on llvm/lib/CAS/HierarchicalTreeBuilder.cpp:
```diff
@@ -257,3 +257,32 @@ Expected<TreeRef> HierarchicalTreeBuilder::create(CASDB &CAS) {
 
+Expected<TreeRef> HierarchicalTreeBuilder::clone(CASDB &Src, TreeRef Root,
+                                                 CASDB &Dst) {
+  HierarchicalTreeBuilder Builder;
```
A bit funny to use a tree builder here, since the builder has to do a bunch of work to figure out how things are nested after the fact.

I think this is a better framework for cloning:
- Create a map from old to new ID.
- Do a topological sort of the DAG rooted by the input (arbitrary object, not necessarily a tree), using a DFS.
- Build objects using the resulting post-order traversal (POT).

We can add peepholes to clone leaf objects during the DFS to avoid having to visit them twice. But that could land later.

dexonsmith replied:
(I don't think that ends up being very complicated code...)

cachemeifyoucan committed:
Rename response-file type and extension

dexonsmith commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -4486,11 +4487,66 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+    }
+    return;
+  }
+
```
You should split the function here and send in the `const JobAction &JA` that you actually want to be used. Then you don't need to change all the `JA` things below.

dexonsmith replied:
Or, somehow find a way to avoid changing every reference to `JA`. We'll get too many build failures and merge conflicts with this the patch looks right now.

E.g., another option is to avoid introducing the `JA` variable until you know what you want it to be set to.

cachemeifyoucan replied:
I have a better solution now. Updating.

dexonsmith replied:
Nice! This looks way better.

dexonsmith commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -4845,7 +4904,7 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
 
     C.addCommand(std::make_unique<Command>(
-        JA, *this, ResponseFileSupport::AtFileUTF8(), D.getClangProgramPath(),
+        Job, *this, ResponseFileSupport::AtFileUTF8(), D.getClangProgramPath(),
```
Why is this different from the others?

cachemeifyoucan replied:
This is intentional. The command creates should be associated with the original job requested, not the overwritten Job. The difference is simply for Depscan job, the command created here is associated with Depscan job, not a compile job.

dexonsmith reviewed (changes requested):
Two high-level comments:
- Need to find a way to avoid the pervasive out-of-tree changes related to `JA` changing from a reference to a pointer. Either introduce the variable later when it can be a reference, or split functions, or something.
- I reviewed the `HierarchicalTreeBuilder::clone` function, but I don't understand why it's in this PR... and it doesn't seem like a good fit for `HierarchicalTreeBuilder`.

cachemeifyoucan commented:
> * I reviewed the `HierarchicalTreeBuilder::clone` function, but I don't understand why it's in this PR... and it doesn't seem like a good fit for `HierarchicalTreeBuilder`.

That is because I totally pushed from the wrong branch. Ignore that.

cachemeifyoucan commented:
@swift-ci please test

dexonsmith commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -4486,11 +4489,66 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+    InputName += II.getFilename();
+    CmdArgs.push_back(C.getArgs().MakeArgString(InputName));
+
+    // FIXME: Overwrite the output filename. To fix this, we need to somehow
+    // route the final output type into Depscan ConstructJob.
+    if (Output.isFilename()) {
+      CmdArgs.push_back("-o");
+      CmdArgs.push_back(Output.getFilename());
+    }
```
Can this be removed now, since the `-cc1depscan` is getting the `-o`?

cachemeifyoucan replied:
`-cc1depscan` doesn't get the correct output file for cc1 argument so this `-o` is passed after response file with the correct output filename. `FIXME` explained how to fix it. It is not hard but need to thread another argument through lots of functions. This is a small overwrite so I think it should be fine for now.

dexonsmith replied:
Okay; the FIXME wasn't clear to me since it doesn't describe the problem. I assumed this was left over from before you'd added the 'combine' step.

I suggest something like this:
```
// FIXME: The -cc1depscan job should include the -o for the -cc1 job that follows
// it in -cc1-args, but it doesn't currently have access to that filename. To fix this,
// we need to somehow route the final output type into Depscan ConstructJob.
```

dexonsmith commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -4486,11 +4489,66 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+      CmdArgs.push_back("-o");
+      CmdArgs.push_back(Output.getFilename());
+    }
+    const char *Exec = D.getClangProgramPath();
+    if (D.CC1Main && !D.CCGenDiagnostics) {
```
I'm fine with the code duplication since it's pretty short, but please add a FIXME for outlining into a function somehow when upstreaming.

cachemeifyoucan replied:
This is not code duplication. This is a new early return.

dexonsmith replied:
I know the early return is new. The logic that takes place starting here, before the return, is non-trivial and essentially identical to the code later in the function where it was copied from. Properly factored, there'd be a lambda function for this `if` block and the call to `getClangProgramPath()`.

cachemeifyoucan replied:
Sure. I will add a note.

cachemeifyoucan committed:
Address review feedback for Depscan driver

cachemeifyoucan commented:
@dexonsmith Going back to comments about performance, I implement that hack and that doesn't make a difference in performance at all. It is probably not file I/O but probably all the logics in the driver and command line option parsing.

Looking at min time, it is about: 13.9s vs 13.4s which is 3-4% for fully cached build.

dexonsmith commented:
> @dexonsmith Going back to comments about performance, I implement that hack and that doesn't make a difference in performance at all. It is probably not file I/O but probably all the logics in the driver and command line option parsing.
> 
> Looking at min time, it is about: 13.9s vs 13.4s which is 3-4% for fully cached build.

I think this is a better model, and we can probably find a way to optimize away the extra work.

Part of what makes me think so is that I *think* there's headroom in `CompilerInvocation` <-> `-cc1` serialization/deserialization (basically, anything in clang/include/clang/Driver/Options.td that has `Marshalling`, `Normalizer`, or `Denormalizer`). It has not be profiled and optimized and it's not on any hot paths (only really used in explicit modules, and here). I remember when I was helping with reviews, there were some wasteful things that we decided were good enough until they became observable as a performance bottleneck.

@cachemeifyoucan, does your intuition agree with mine, that the bottleneck is in extra `-cc1` / CompilerInvocation serialization?
- If so, can you think of any ways to confirm?
- Else, where do you think it is? 

@benlangmuir @akyrtzi , WDYT? Is 3-4% an okay regression for a cached build?

benlangmuir commented:
> WDYT? Is 3-4% an okay regression for a cached build?

I'm not concerned as long as we think it's reasonably fixable.  Would be good to confirm one way or another first.

akyrtzi commented:
Also make sure there's a FIXME and associated radar so we don't lose track.

cachemeifyoucan commented:
I benchmarked one invocation in Instruments (add a loop around depscan_main) to see if I can spot the extra overhead in OptTable or Arg serialization. It turns out that I forgot to remove the output file when I implement the hacky response file passed in memory so even I didn't write anything to the response file, it is still created and removed. And from the time profiler, that is all the measurable overhead from doing it this way.

I can add a FIXME to document the overhead for now. Maybe we can teach clang to expand response file from CAS `@llvmcas://HASH` and make a default in memory CAS for clang driver to pass intermediate file using that CAS when everything is in memory?

cachemeifyoucan committed:
Add comments about the overhead of depscan responsefile

dexonsmith commented:
> I benchmarked one invocation in Instruments (add a loop around depscan_main) to see if I can spot the extra overhead in OptTable or Arg serialization. It turns out that I forgot to remove the output file when I implement the hacky response file passed in memory so even I didn't write anything to the response file, it is still created and removed. And from the time profiler, that is all the measurable overhead from doing it this way.

Great news!

> I can add a FIXME to document the overhead for now. Maybe we can teach clang to expand response file from CAS `@llvmcas://HASH` and make a default in memory CAS for clang driver to pass intermediate file using that CAS when everything is in memory?

There isn't really a hash we can put there, since we don't know what the hash will be. We'd need to introduce the idea of a name. If we had the concept of a naming service, we could share a naming service between both actions.

To be clear, by "naming service" I mean something like what git-branch manages, which is a read-write map associating a name with an arbitrary object.

Really what we want is a new way to pass an `InputInfo` from one job to the next that doesn't require writing out a file.
- If the jobs are in the same process, should work regardless of CAS config, since we can write to the CAS.
- If the jobs are in a different process, write the temp file to disk unless there's an appropriate persistent naming service to use.
- If using `-save-temps`, you still write to disk (but no need to read from there if the above is sufficient for reading). 

Without a naming service (per se), another approach could be:
- Configure an OutputBackend for the writing action that captures just the main output instead of writing it to disk.
- Use a VFS for the reading action that adds the created tree as overlay.
- This would take some work since I don't think command-line parsing takes a VFS for accessing response files.

Yet another approach is to set up a pipeline (although we'd probably want `-###` output to lie and use a temp file):
- The writing action thinks it's writing to `-o -`.
- The reading action thinks it's reading from `@-`.

In all of these latter approaches, we'd want the "in-memory" `-cc1` and `-cc1depscan` invocations to take an extra hook from the driver as an argument (e.g., the in-memory CAS and naming service to share). Maybe the necessary context could be constructed during `combine...()`, saved in the `JobAction`, and passed through to the `CC1Command`s.

dexonsmith commented:
> > I benchmarked one invocation in Instruments (add a loop around depscan_main) to see if I can spot the extra overhead in OptTable or Arg serialization. It turns out that I forgot to remove the output file when I implement the hacky response file passed in memory so even I didn't write anything to the response file, it is still created and removed. And from the time profiler, that is all the measurable overhead from doing it this way.
> 
> Great news!

BTW, this also seems to confirm that there's headroom available by avoiding writing *any* object files in a cached build.
- Maybe `clone_file()` hacks would be faster, although they'll still need file creation.
- Maybe the `tablegen` trick that avoids writing an output if it's bit-for-bit identical to the existing file would be useful?

... or maybe we just need/want a naming service and the overhead will stick around until we can thread this through all the way.

dexonsmith reviewed (approved):
Just noticed you added the FIXME already. This LGTM!
